pkg/networkservice/common/discover/server_test.go

```diff
@@ -21,133 +21,237 @@ import (
 	"context"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
+
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 
 	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 	"github.com/networkservicemesh/api/pkg/api/registry"
-	"github.com/stretchr/testify/assert"
 	"google.golang.org/grpc"
 
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/discover"
 )
 
+type contextKeyType string
+
+const (
+	testDataKey contextKeyType = "testData"
+)
+
+type TestData struct {
+	endpoints []*registry.NetworkServiceEndpoint
+}
+
+func withTestData(parent context.Context, testData *TestData) context.Context {
+	if parent == nil {
+		parent = context.TODO()
+	}
+	return context.WithValue(parent, testDataKey, testData)
+}
+
+func testData(ctx context.Context) *TestData {
+	if rv, ok := ctx.Value(testDataKey).(*TestData); ok {
+		return rv
+	}
+	return nil
+}
+
 type testNetworkServiceServer struct {
-	t    *testing.T
-	want []*registry.NetworkServiceEndpoint
 }
 
 func (s testNetworkServiceServer) Request(ctx context.Context, in *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
```

Comment:
I think it become more complicated, then it supposed it be. For such scenarios we should have a Network Service accepting function, and function should do assertion inplace.

 Like 
```go
request := ...
server := next.NewNetworkServiceServer(discover.NewServer(discoveryClient), &AssertNetworkService{
Before: func(req networkservice.NetworkServiceRequest ) {
  // do required checks.
},
After: func(conn *connection.Connection)  {
 // do checks on connection
}
})
```

Thoughts?

Reply:
You should be able to accomplish checking after with

[checkcontext](https://github.com/networkservicemesh/sdk/tree/master/pkg/networkservice/utils/checks/checkcontext)

```golang
testServer := chain.NewNetworkServiceServer(
   discover.NewServer(discoveryClient),
   checkcontext.NewServer(func(t *testing.T, ctx context.Context),
}
```

If you need to check something about the after you can use: [checkrequest](https://github.com/networkservicemesh/sdk/tree/master/pkg/networkservice/utils/checks/checkrequest) or if you need to check the connection on return you can use [checkconnection](https://github.com/networkservicemesh/sdk/tree/master/pkg/networkservice/utils/checks/checkconnection).  

Finally, if you need to check something about the context *on* return (rather than after) you can use [checkcontextonreturn](https://github.com/networkservicemesh/sdk/tree/master/pkg/networkservice/utils/checks/checkcontextonreturn)